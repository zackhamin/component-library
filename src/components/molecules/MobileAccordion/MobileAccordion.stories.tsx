import { Button } from 'Atoms/Button';
import { Heading } from 'Atoms/Heading';
import { Story } from '@storybook/react';
import { MobileAccordion, MobileAccordionProps } from './MobileAccordion';

export default {
	title: 'molecules/MobileAccordion',
	component: MobileAccordion,
};

const content = (
	<>
		<Heading level={3} size="1.6rem">
			Mobile accordion content
		</Heading>
		<Button size="small">Button</Button>
	</>
);

const Template: Story<MobileAccordionProps> = (args) => (
	<MobileAccordion {...args} />
);

const defaultProps = {
	title: 'standard title as a string',
	content: content,
	expanded: false,
	testId: 'MobileAccordion',
	id: '',
	className: '',
} as MobileAccordionProps;

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Expanded = Template.bind({});
Expanded.args = {
	...defaultProps,
	expanded: true,
};

export const WithFooterContent = Template.bind({});
WithFooterContent.args = {
	...defaultProps,
	content: <div>Content</div>,
	footer: <Button size="small">Footer Content</Button>,
};
