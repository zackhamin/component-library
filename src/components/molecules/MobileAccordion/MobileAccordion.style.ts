import styled from 'styled-components';
import { baseStyles, colors } from 'Styles';

export const StyledMobileAccordion = styled.div`
	${baseStyles};
	display: flex;
	flex-direction: column;
	overflow: hidden;
	box-shadow: 0px 8px 24px rgba(51, 51, 51, 0.16);
	border-radius: 8px 8px 0px 0px;
	padding: 2.4rem;

	@media screen and (min-width: 768px) {
		border-radius: 8px;
	}
`;

export const StyledMobileAccordionTitle = styled.div`
	display: flex;
	align-items: center;
	justify-content: space-between;
	background-color: ${colors.transparent};
	padding-bottom: 0.4rem;
	border: 0;
	margin: 0.125rem;
	position: relative;

	h2 {
		margin: 0;
		color: ${colors.offBlack};
		font-size: 1.6rem;
		line-height: 160%;
	}
`;

export const StyledMobileAccordionIcon = styled.button`
	width: 1.6rem;
	height: 1.6rem;
	background: none;
	border: none;
	color: ${colors.offBlack};
	padding: 0;
	position: static;
	cursor: pointer;

	&:before {
		content: '';
		display: block;
		position: absolute;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
	}
`;

export const StyledMobileAccordionContent = styled.div<{
	isExpanded?: boolean;
}>`
	height: 0;
	overflow: auto;
	visibility: hidden;
	padding: 0;

	-ms-overflow-style: none;
	scrollbar-width: none;

	&::-webkit-scrollbar {
		display: none;
	}

	${({ isExpanded }) =>
		isExpanded &&
		`padding: 0.4rem 0;
      height: 100%;  
	  display:block;
	  visibility: visible;
      `}
`;

export const StyledMobileAccordionFooter = styled.div`
	padding-top: 0.8rem;
`;
