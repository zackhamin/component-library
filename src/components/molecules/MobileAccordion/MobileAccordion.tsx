import { Heading } from 'Atoms/Heading';
import { isEventValid } from 'Utils/utils';
import { IconMinus, IconPlus } from 'Atoms/Icons';
import React, { ReactNode, useEffect, useState } from 'react';
import {
	StyledMobileAccordion,
	StyledMobileAccordionContent,
	StyledMobileAccordionFooter,
	StyledMobileAccordionIcon,
	StyledMobileAccordionTitle,
} from './MobileAccordion.style';

export type MobileAccordionProps = React.HTMLAttributes<HTMLDivElement> & {
	title: string;
	content?: ReactNode;
	expanded?: boolean;
	maxScreenWidth?: number;
	footer?: ReactNode;
};

export const MobileAccordion = ({
	title,
	content,
	expanded,
	maxScreenWidth = 768,
	footer,
	...props
}: MobileAccordionProps) => {
	if (expanded != true) {
		expanded = false;
	}
	const [isMobile, setIsMobile] = useState<boolean>(
		window.innerWidth <= maxScreenWidth
	);
	const [isExpanded, setIsExpanded] = useState<boolean>(expanded);

	useEffect(() => {
		function handleResize() {
			setIsMobile(window.innerWidth <= maxScreenWidth);
		}

		window.addEventListener('resize', handleResize);

		return () => window.removeEventListener('resize', handleResize);
	});

	const toggleIsExpanded = (
		event: React.MouseEvent<HTMLButtonElement, MouseEvent>
	) => {
		isEventValid(event) && setIsExpanded((prevIsExpanded) => !prevIsExpanded);
	};

	return (
		<StyledMobileAccordion {...props}>
			<StyledMobileAccordionTitle>
				<Heading level={2} size="1.6rem">
					{title}
				</Heading>
				{isMobile ? (
					<StyledMobileAccordionIcon
						aria-expanded={isExpanded}
						aria-label={title}
						onClick={toggleIsExpanded}
					>
						{isExpanded ? (
							<IconMinus fontSize="1.6rem" data-testid="MINUS" />
						) : (
							<IconPlus fontSize="1.6rem" data-testid="PLUS" />
						)}
					</StyledMobileAccordionIcon>
				) : null}
			</StyledMobileAccordionTitle>

			<StyledMobileAccordionContent isExpanded={!isMobile || isExpanded}>
				{content}
			</StyledMobileAccordionContent>

			{footer && (
				<StyledMobileAccordionFooter>{footer}</StyledMobileAccordionFooter>
			)}
		</StyledMobileAccordion>
	);
};
