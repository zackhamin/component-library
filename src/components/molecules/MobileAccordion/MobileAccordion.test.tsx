import { act } from 'react-dom/test-utils';
import { MobileAccordion } from './MobileAccordion';
import { fireEvent, render } from '@testing-library/react';

describe('MobileAccordion', () => {
	const defaultProps = {
		title: 'Mobile accordion title',
		content: 'Mobile accordion content',
	};

	test('renders without errors and matches snapshot', () => {
		const {
			container: { firstChild },
		} = render(<MobileAccordion {...defaultProps} />);

		expect(firstChild).toMatchSnapshot();
	});

	describe('Desktop view', () => {
		test('should display title and content', () => {
			const { getByText } = render(<MobileAccordion {...defaultProps} />);
			const accordionTitle = getByText(/title/i);
			const accordionContent = getByText(/content/i);
			expect(accordionTitle).toBeVisible();
			expect(accordionContent).toBeVisible();
		});

		test('Should not have a button to show content', () => {
			const { queryByRole } = render(<MobileAccordion {...defaultProps} />);
			const button = queryByRole('button');
			expect(button).toBeNull();
		});
	});

	describe('Mobile view', () => {
		test('should display title, button and icon', () => {
			act(() => {
				global.innerWidth = 500;
				global.dispatchEvent(new Event('resize'));
			});
			const { getByRole, getByTestId, getByText, queryByText } = render(
				<MobileAccordion {...defaultProps} />
			);
			const accordionTitle = getByText(/title/i);
			const accordionContent = queryByText(/content/i);
			const button = getByRole('button');
			const plusIcon = getByTestId('PLUS');

			expect(accordionTitle).toBeVisible();
			expect(button).toBeVisible();
			expect(plusIcon).toBeVisible();
			expect(accordionContent).not.toBeVisible();
		});

		test('should show content and updated icon on click', () => {
			act(() => {
				global.innerWidth = 500;
				global.dispatchEvent(new Event('resize'));
			});
			const { getByRole, getByTestId, getByText } = render(
				<MobileAccordion {...defaultProps} />
			);
			const button = getByRole('button');
			fireEvent.click(button);

			const accordionContent = getByText(/content/i);
			const minusIcon = getByTestId('MINUS');
			expect(accordionContent).toBeVisible();
			expect(minusIcon).toBeVisible();
		});

		test('should show content when expanded is true', () => {
			act(() => {
				global.innerWidth = 500;
				global.dispatchEvent(new Event('resize'));
			});
			const { getByText } = render(
				<MobileAccordion {...defaultProps} expanded={true} />
			);

			const accordionContent = getByText(/content/i);
			expect(accordionContent).toBeVisible();
		});
	});
});
