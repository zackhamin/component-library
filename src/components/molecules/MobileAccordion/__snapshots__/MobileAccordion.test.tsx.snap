// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileAccordion renders without errors and matches snapshot 1`] = `
.c2 {
  margin: 0;
  font-family: BrandonGrotesque-Bold,Arial,Helvetica,sans-serif;
  color: #333333;
  font-size: 1.6rem;
  line-height: 160%;
  font-weight: 450;
}

.c0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.2rem;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  box-shadow: 0px 8px 24px rgba(51,51,51,0.16);
  border-radius: 8px 8px 0px 0px;
  padding: 2.4rem;
}

.c0:focus:not(:focus-visible) {
  outline: none;
}

.c0:focus-visible {
  outline: 0.125rem solid #212529;
}

.c0 h1 {
  color: #333;
  font-size: 3.2rem;
  line-height: 1.75;
  margin: 0 0 1rem;
}

.c0 h2 {
  color: #00594c;
  font-size: 2.4rem;
  line-height: 1.75;
  margin: 0 0 1rem;
}

.c0 p {
  font-size: 1.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background-color: Transparent;
  padding-bottom: 0.4rem;
  border: 0;
  margin: 0.125rem;
  position: relative;
}

.c1 h2 {
  margin: 0;
  color: #333333;
  font-size: 1.6rem;
  line-height: 160%;
}

.c3 {
  height: 0;
  overflow: auto;
  visibility: hidden;
  padding: 0;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  padding: 0.4rem 0;
  height: 100%;
  display: block;
  visibility: visible;
}

.c3::-webkit-scrollbar {
  display: none;
}

@media screen and (min-width:768px) {
  .c0 {
    border-radius: 8px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <h2
      class="c2"
    >
      Mobile accordion title
    </h2>
  </div>
  <div
    class="c3"
  >
    Mobile accordion content
  </div>
</div>
`;
